### PR TITLE
feat: custom slippage improvements for UI and cowswap swapper

### LINKED ISSUE #5779: Handle quotes that dont support custom slippage
### Overview

Some types of swappers (thorchain streaming swaps, cowswap) dont support the ability to specify the slippage when getting a quote, and we need a way to propagate this to users.

Implement a UI change that shows users what quotes are affected, using the attached mockup.

### References and additional details

- Also show another alert above the preview button IF that quote is selected.
- Each quote will need to show the slippage, regardless of whether custom slippage was specified (for consistency and utility)
- If the slippage is different than than the custom value, we show this warning state for the slippage
- If custom slippage was not specified, dont show any warnings

### Acceptance Criteria

- Slippage dispayed on all quotes
- Quotes ignoring the custom slippage (thorchain streaming swaps, cowswap) show warning icon with tolltip
- Alert appears above the preview button when a quote ignoring custom slippage is selected

### Need By Date

_No response_

### Screenshots/Mockups

image

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Some types of swappers (thorchain streaming swaps) dont support the ability to specify the slippage when getting a quote, and we need a way to propagate this to users. Other swappers (0x) only provide slippage protection for certain pairs.

This PR implements a UI change that shows users what quotes are affected.

- Includes custom slippage capability for cowsap
- 0x only provides slippage protection for specific pairs, so this PR addresses that
- Fixes trade execution UI showing the default slippage instead of custom slippage

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #5779

## Risk

- Risk of broken trades
- Risk of incorrect custom slippage for cow swap

## Testing

- Check trades are working as expected
- Check cowswap custom slippage is correct

### Engineering

Carefully check the implementation of custom slippage for cowswap.

## Screenshots (if applicable)

Cowswap default slippage value - see matching payload
image

Cowswap custom slippage value - see matching payload
image

Completed cow trade with custom slippage (note bottom right of UI 
image

Video showing slippage UI behaviour

https://github.com/shapeshift/web/assets/125113430/fe60fdef-092d-4f0b-adcd-15aa80470e63

### CODE REVIEW

**Comment by woodenfurniture** [comment]
Current dependencies on/for this PR:
* `develop`
  * **PR #5908** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5908?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/5908?utm_source=stack-comment).

File: src/components/MultiHopTrade/components/TradeInput/components/TradeQuotes/TradeQuote.tsx [hunk 1]
```diff
@@ -292,6 +331,17 @@ export const TradeQuoteLoaded: FC<TradeQuoteProps> = ({
...
+              {totalEstimatedExecutionTimeMs !== undefined && totalEstimatedExecutionTimeMs > 0 && (
+                <Skeleton isLoaded={!isLoading}>
+                  <Flex gap={2} alignItems='center'>
+                    <RawText color='text.subtle'>
+                      <FaRegClock />
+                    </RawText>
+                    {prettyMilliseconds(totalEstimatedExecutionTimeMs)}
+                  </Flex>
+                </Skeleton>
+              )}
```
Comment (woodenfurniture) [inline src/components/MultiHopTrade/components/TradeInput/components/TradeQuotes/TradeQuote.tsx:344]: moved this below so UI shifts layout less often (easier on eyes)

File: src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx [hunk 2]
```diff
@@ -596,10 +594,7 @@ export const TradeInput = memo(() => {
                     </FadeTransition>
                   )}
                 </AnimatePresence>
-
-                {(activeSwapperSupportsSlippage || sortedQuotes.length === 0) && (
-                  <SlippagePopover />
-                )}
```
Comment (woodenfurniture) [inline src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx:602]: all swappers now support custom slippage

File: src/components/MultiHopTrade/components/MultiHopTradeConfirm/components/Hop.tsx [hunk 3]
```diff
@@ -115,11 +116,6 @@ export const Hop = ({
     }
   }, [hopExecutionState, hopIndex])
 
-  const slippageDecimalPercentage = useMemo(
-    () => getDefaultSlippageDecimalPercentageForSwapper(swapperName),
-    [swapperName],
-  )
-
```
Comment (woodenfurniture) [inline src/components/MultiHopTrade/components/MultiHopTradeConfirm/components/Hop.tsx:122]: this was clearly a bug

File: src/lib/swapper/swappers/CowSwapper/utils/helpers/helpers.ts [hunk 4]
```diff
@@ -230,10 +234,12 @@ const generateAppDataFromDoc = async (
...
 const metadataApi = new MetadataApi()
 // See https://api.cow.fi/docs/#/default/post_api_v1_quote / https://github.com/cowprotocol/app-data
-export const getFullAppData = async () => {
+export const getFullAppData = async (slippageTolerancePercentage: string) => {
   const APP_CODE = 'shapeshift'
   const orderClass: OrderClass = { orderClass: 'market' }
-  const quote = { slippageBips: '50' }
+  const quote = {
+    slippageBips: convertDecimalPercentageToBasisPoints(slippageTolerancePercentage).toString(),
+  }
```
Comment (woodenfurniture) [inline src/lib/swapper/swappers/CowSwapper/utils/helpers/helpers.ts:242]: credit to @gomesalexandre for setting the foundation for this to be possible

File: src/components/MultiHopTrade/components/TradeInput/components/TradeQuotes/TradeQuote.tsx [hunk 5]
```diff
@@ -193,6 +202,46 @@ export const TradeQuoteLoaded: FC<TradeQuoteProps> = ({
     [quote?.steps],
   )
 
+  const slippage = useMemo(() => {
+    if (!quote) return
+
+    const isError =
+      userSlippageDecimal !== undefined && quote.slippageTolerancePercentage !== userSlippageDecimal
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/TradeInput/components/TradeQuotes/TradeQuote.tsx:209]: Non-blocking, but can we use a consistent vernacular for both of these? i.e `slippageDecimal` or `slipperPercentage` (perhaps `percentageDecimal`).
Reading this is a bit confusing

Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/TradeInput/components/TradeQuotes/TradeQuote.tsx:209]: Also, can we use a different terminology than error here? Not being able to set custom slippage is not an error, it's part of the flow for swappers that don't support it, so this adds to the confusion

File: src/lib/swapper/swappers/ThorchainSwapper/utils/getL1quote.ts [hunk 6]
```diff
@@ -218,6 +218,7 @@ export const getL1quote = async (
               potentialAffiliateBps,
               isStreaming,
               recommendedMinimumCryptoBaseUnit,
+              slippageTolerancePercentage: isStreaming ? undefined : slippageTolerancePercentage,
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/getL1quote.ts:221]: 🎉

**Review by gomesalexandre — COMMENTED** [review]
Tested locally. Conceptual stamp except for one question.

### 1inch

- Default slippage

image

- Custom slippage is honored at final quote time

image
image
image

### CoW

- Default slippage

image

- Custom slippage is honored

image
image
image

According to this implementation:

https://github.com/cowprotocol/cowswap/blob/071ce71b22d25a7767547f39254a5845132622a4/libs/common-utils/src/misc.ts#L176-L182

100 indeed is correct as 1% slippage ✅ 
image

### THOR non-streaming swaps

- Default slippage 


- Custom slippage
 
image
image

@woodenfurniture the default 1% doesn't seem to be reflected at quote time here, though I believe this is on purpose because of our implementation?

### Streaming swaps

- Default slippage bps

image
image

- Custom slippage

image
image
image

Custom slippage is unable to get set for streaming swaps, and auto bps are used as they should ✅

**Comment by woodenfurniture** [comment]
> @woodenfurniture the default 1% doesn't seem to be reflected at quote time here, though I believe this is on purpose because of our implementation, as we use limit over slippage? Memo for reference: =:THOR.RUNE:thor125dwsa39yeylqc7pn59l079dur502nsleyrgup:194410385:ss:0

correct, we're manually implementing slippage through the limit mechanism here:
`src/lib/swapper/swappers/ThorchainSwapper/utils/addSlippageToMemo.ts`
image

File: src/components/MultiHopTrade/components/TradeInput/components/TradeQuotes/TradeQuote.tsx (hunk 5 above)
Comment (woodenfurniture) [reply src/components/MultiHopTrade/components/TradeInput/components/TradeQuotes/TradeQuote.tsx]: yeah agreed, I've pushed a commit to align things a bit better 8996988ec0cb8bf1fab291908c6cb4a3938141fc

Comment (woodenfurniture) [reply src/components/MultiHopTrade/components/TradeInput/components/TradeQuotes/TradeQuote.tsx]: also a88fa6bd01b8a1d81fb6baf3a448fc14adb6e09b

**Review by gomesalexandre — APPROVED** [review]
Akschual stamp after q. has been answered.

**Pull request merged**